### PR TITLE
refactor any parser classes and update testcases

### PR DESCRIPTION
**Pull request by Sdddell**
## Description
- Refactor sync anyparser classes, and async anyparser classes.

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [x] Code refactoring
- [ ] Performance improvement

## How Has This Been Tested?
Local unittest and test through requests.

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Additional Notes

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Refactored `AnyParser` to replace `SyncParser` with multiple specialized parser classes for modularity and maintainability.
- Introduced `BasePostProcessor` and its subclasses for modular response processing in `AsyncParser`.
- Simplified `BaseParser` by removing the `_setup_endpoints` method.
- Added new specialized sync parser classes for parsing and extracting specific data types (e.g., PII, tables, key-value pairs).
- Updated test outputs for DOCX, PDF, PNG, and PPTX files to include table markers and improve formatting.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>any_parser.py</strong><dd><code>Refactor `AnyParser` to use specialized parser classes.</code>&nbsp; &nbsp; </dd></summary>
<hr>

any_parser/any_parser.py

<li>Refactored <code>AnyParser</code> class to replace <code>SyncParser</code> with multiple <br>specialized parsers.<br> <li> Updated methods to use new parser classes for specific tasks like <br>parsing, extracting PII, tables, and key-value pairs.<br> <li> Simplified response handling by removing redundant error-checking <br>logic.<br> <li> Updated base URL for API endpoints.<br>

</details>

  </td>
  <td><a href="https://github.com/CambioML/any-parser/pull/70/files#diff-b580101abf5c2c14a05ba9e6a763e462c65136ba8e884488ddcd93561600346d">+36/-92</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>async_parser.py</strong><dd><code>Add modular post-processing for async parser responses.</code>&nbsp; &nbsp; </dd></summary>
<hr>

any_parser/async_parser.py

<li>Introduced <code>BasePostProcessor</code> and its subclasses for modular response <br>processing.<br> <li> Refactored <code>AsyncParser</code> to use the new post-processing chain for <br>handling JSON responses.<br> <li> Added a <code>handle_async_response</code> method for streamlined response <br>handling.<br>

</details>

  </td>
  <td><a href="https://github.com/CambioML/any-parser/pull/70/files#diff-62460a4a6e9b6b32a300b4f8bfb988f895bf3efb03710ee76207ec1ef5029f16">+58/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>base_parser.py</strong><dd><code>Simplify `BaseParser` by removing endpoint setup.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

any_parser/base_parser.py

<li>Removed <code>_setup_endpoints</code> method from <code>BaseParser</code>.<br> <li> Simplified initialization by removing endpoint setup logic.<br>

</details>

  </td>
  <td><a href="https://github.com/CambioML/any-parser/pull/70/files#diff-6917b7633fc59d87b01d902fcd184f5f2f8a721014c1e24f9b14b1647d9b2a4b">+0/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sync_parser.py</strong><dd><code>Modularize sync parsing with specialized parser classes.</code>&nbsp; </dd></summary>
<hr>

any_parser/sync_parser.py

<li>Replaced <code>SyncParser</code> with multiple specialized parser classes <br>inheriting from <code>BaseSyncParser</code>.<br> <li> Added implementations for parsing and extracting specific data types <br>(e.g., PII, tables, key-value pairs).<br> <li> Modularized sync parsing logic for better maintainability.<br>

</details>

  </td>
  <td><a href="https://github.com/CambioML/any-parser/pull/70/files#diff-41dcc98722eb8170868f074fba55be15a6995aef0b1366cf33be9535e3faa8f0">+161/-10</a></td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>correct_docx_output.txt</strong><dd><code>Update DOCX test output with table markers.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/outputs/correct_docx_output.txt

<li>Added table markers (<code><|table_start|></code> and <code><|table_end|></code>) to improve output structure.<br> <li> Adjusted formatting for better readability.<br>

</details>

  </td>
  <td><a href="https://github.com/CambioML/any-parser/pull/70/files#diff-eabc24c972cf72fc23a07dc5f45f916640999286f6e56abf825c88acd3295950">+5/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>correct_pdf_output.txt</strong><dd><code>Update PDF test output with table markers.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/outputs/correct_pdf_output.txt

<li>Added table markers (<code><|table_start|></code> and <code><|table_end|></code>) to improve output structure.<br> <li> Simplified and reformatted content for consistency.<br>

</details>

  </td>
  <td><a href="https://github.com/CambioML/any-parser/pull/70/files#diff-b83d1b767cf9e3fad3963cd24861077247758154164fbcad712618c6223881c8">+112/-131</a></td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>correct_png_output.txt</strong><dd><code>Update PNG test output with table markers.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/outputs/correct_png_output.txt

<li>Added table markers (<code><|table_start|></code> and <code><|table_end|></code>) to improve output structure.<br> <li> Adjusted formatting for better readability.<br>

</details>

  </td>
  <td><a href="https://github.com/CambioML/any-parser/pull/70/files#diff-ce565eb5ec25167834e159aea9f80595f57003305e95b8470ee7c27ac8570fc6">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>correct_pptx_output.txt</strong><dd><code>Update PPTX test output with table markers.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/outputs/correct_pptx_output.txt

<li>Added table markers (<code><|table_start|></code> and <code><|table_end|></code>) to improve output structure.<br> <li> Adjusted formatting for better readability.<br>

</details>

  </td>
  <td><a href="https://github.com/CambioML/any-parser/pull/70/files#diff-f481a787eb177a0402b4848b0e4c543d196ec10dfd6f6e43d5e62ffd96f5c0fa">+4/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by lingjiekong** [comment]
> ### **User description**
> ## Description
> * Refactor sync anyparser classes, and async anyparser classes.
> 
> ## Type of Change
> * [ ]  Bug fix (non-breaking change which fixes an issue)
> * [ ]  New feature (non-breaking change which adds functionality)
> * [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
> * [ ]  Documentation update
> * [x]  Code refactoring
> * [ ]  Performance improvement
> 
> ## How Has This Been Tested?
> Local unittest and test through requests.
> 
> ## Checklist
> * [x]  My code follows the project's style guidelines
> * [x]  I have performed a self-review of my own code
> * [x]  I have commented my code, particularly in hard-to-understand areas
> * [ ]  I have made corresponding changes to the documentation
> * [x]  My changes generate no new warnings
> * [x]  I have added tests that prove my fix is effective or that my feature works
> * [x]  New and existing unit tests pass locally with my changes
> 
> ## Additional Notes
> ### **PR Type**
> Enhancement, Tests
> 
> ### **Description**
> * Refactored `AnyParser` to replace `SyncParser` with multiple specialized parser classes for modularity and maintainability.
> * Introduced `BasePostProcessor` and its subclasses for modular response processing in `AsyncParser`.
> * Simplified `BaseParser` by removing the `_setup_endpoints` method.
> * Added new specialized sync parser classes for parsing and extracting specific data types (e.g., PII, tables, key-value pairs).
> * Updated test outputs for DOCX, PDF, PNG, and PPTX files to include table markers and improve formatting.
> 
> ### **Changes walkthrough** 📝
> Relevant files
> **Enhancement**	
> **any_parser.py**
> `` Refactor `AnyParser` to use specialized parser classes. ``   
> [+36/-92](https://github.com/CambioML/any-parser/pull/70/files#diff-b580101abf5c2c14a05ba9e6a763e462c65136ba8e884488ddcd93561600346d) 
> **async_parser.py**
> `Add modular post-processing for async parser responses.`   
> [+58/-1](https://github.com/CambioML/any-parser/pull/70/files#diff-62460a4a6e9b6b32a300b4f8bfb988f895bf3efb03710ee76207ec1ef5029f16)   
> **base_parser.py**
> `` Simplify `BaseParser` by removing endpoint setup. ``               
> [+0/-5](https://github.com/CambioML/any-parser/pull/70/files#diff-6917b7633fc59d87b01d902fcd184f5f2f8a721014c1e24f9b14b1647d9b2a4b)     
> **sync_parser.py**
> `Modularize sync parsing with specialized parser classes.` 
> [+161/-10](https://github.com/CambioML/any-parser/pull/70/files#diff-41dcc98722eb8170868f074fba55be15a6995aef0b1366cf33be9535e3faa8f0)
> **Tests**	
> **correct_docx_output.txt**
> `Update DOCX test output with table markers.`                           
> [+5/-3](https://github.com/CambioML/any-parser/pull/70/files#diff-eabc24c972cf72fc23a07dc5f45f916640999286f6e56abf825c88acd3295950)     
> **correct_pdf_output.txt**
> `Update PDF test output with table markers.`                             
> [+112/-131](https://github.com/CambioML/any-parser/pull/70/files#diff-b83d1b767cf9e3fad3963cd24861077247758154164fbcad712618c6223881c8)
> **correct_png_output.txt**
> `Update PNG test output with table markers.`                             
> [+2/-2](https://github.com/CambioML/any-parser/pull/70/files#diff-ce565eb5ec25167834e159aea9f80595f57003305e95b8470ee7c27ac8570fc6)     
> **correct_pptx_output.txt**
> `Update PPTX test output with table markers.`                           
> [+4/-2](https://github.com/CambioML/any-parser/pull/70/files#diff-f481a787eb177a0402b4848b0e4c543d196ec10dfd6f6e43d5e62ffd96f5c0fa)     
> > 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

Let's check the repo doc and the mintlify to make sure all docs are up to date, so we have a good practice to update all related documents.

**Review by lingjiekong — APPROVED** [review]
LGTM with comment to address in next PR.

File: any_parser/any_parser.py [hunk 1]
```diff
@@ -141,23 +153,13 @@ def parse(
...
         """
-        response, info = self._sync_parser.get_sync_response(
-            self._sync_parser._sync_parse_url,
-            file_content=file_content,  # type: ignore
-            file_type=file_type,  # type: ignore
+        return self._sync_parse.parse(
+            file_path=file_path,
+            file_content=file_content,
+            file_type=file_type,
             extract_args=extract_args,
```
[the remark this answers is not in the document]
Comment (Sdddell) [reply any_parser/any_parser.py:160]: The parse uses VQAProcessor which also has some extract args (VqaProcessorArgs).

File: any_parser/async_parser.py [hunk 2]
```diff
@@ -58,3 +98,20 @@ def send_async_request(
 
         # If response successful, upload the file
         return upload_file_to_presigned_url(file_content, response)
+
+    def handle_async_response(self, response) -> str:
+        if response is None:
+            return "Error: timeout, no response received"
+        if response.status_code == 202:
+            return ""
```
[the remark this answers is not in the document]
Comment (Sdddell) [reply any_parser/async_parser.py:106]: ACK

**Pull request merged**